docs/code/CrossPlatform.md

```diff
@@ -0,0 +1,179 @@
+# Cross-Platform and Architecture design doc
+
+### Table of contents  <!-- omit in toc -->
+- [Cross-Platform and Architecture design doc](#cross-platform-and-architecture-design-doc)
+	- [1. Why cross-platform/architecture](#1-why-cross-platformarchitecture)
+	- [2. Current status](#2-current-status)
+		- [2.1 Problems](#21-problems)
+		- [2.2 Build](#22-build)
+		- [2.3 Managed Code](#23-managed-code)
+		- [2.4 Native Projects](#24-native-projects)
+		- [2.5 3rd Party Dependencies](#25-3rd-party-dependencies)
+	- [3 Possible Solutions](#3-possible-solutions)
+		- [3.1 Eliminate native components](#31-eliminate-native-components)
+		- [3.2 Rewrite native components to work on other platforms](#32-rewrite-native-components-to-work-on-other-platforms)
+		- [3.3 Rewrite native components to be only managed code](#33-rewrite-native-components-to-be-only-managed-code)
+		- [3.4 Hybrid Between Finding Replacements and Software Fallbacks](#34-hybrid-between-finding-replacements-and-software-fallbacks)
+	- [My Suggestion](#my-suggestion)
+		- [Improving managed fallback through intrinsics](#improving-managed-fallback-through-intrinsics)
+		- [3rd Party Dependencies](#3rd-party-dependencies)
+		- [Helix](#helix)
+		- [Mobile Support](#mobile-support)
+		- [Support Grid](#support-grid)
+
+## 1. Why cross-platform/architecture
+ML.NET is an open-source machine learning framework which makes machine learning accessible to .NET developers with the same code that powers machine learning across many Microsoft products, including Power BI, Windows Defender, and Azure.
+
+ML.NET allows .NET developers to develop/train their own models and infuse custom machine learning into their applications using .NET.
+
+Currently, while .NET is able to run on many different platforms and architectures, ML.NET is only able to run on Windows, Mac, and Linux, either x86 or x64. This excludes many architectures such as Arm, Arm64, M1, and web assembly, places that .NET is currently able to run.
+
+The goal is to enable ML.NET to run everywhere that .NET itself is able to run.
+
+## 2. Current status
+There are several problems complicating us from moving to a fully cross-platform solution. At a high level thes include:
+1. We only build and test for a subset of platforms that .NET supports.
+2. Native components must be explicitly built for additional architectures which we wish to support. This limits our ability to support new platforms without doing work.
+3. Building our native components for new platforms faces challenges due to lack of support for those components dependencies. This limits our ability to support the current set of platforms.
+4. Some of our external dependencies have limited support for .NET's supported platforms.
```

Comment:
There are also some things ML.NET uses internally that don't work well on some platforms, e.g. various components are parallelized but current webassembly targets are single threaded.  It's likely some changes will be necessary to various algorithms to work well in these environments.

Reply:
That is a good point I hadn't though about. Let me add that point as well.